Adds UUID Support

kamronbatman opened the pull request:
Adds UUID version 1 through 5 support.  Based on #970.

kamronbatman commented:
I realize that some of tests are failing because braces and separators are optional.  I'll try and come up with a good way to affect what I wrote above about giving the option for specifying (or not) the grouper, separator, and versions.

kamronbatman commented:
I added proper bracket support, bracket/separator match support and support for options in the form of `{ versions: ['uuidv1','uuidv2','uuidv3','uuidv4','uuidv5'] }`.
If no options are specified then it will validate exactly like the previous GUID check where brackets and separators are optional.

Help with testing would be appreciated!  Thanks!

gelliott181 commented:
In your tests, did you generate those UUIDs or copy a single one and modify its versioning? This comes up because doing so means you may be validating out-of-spec UUIDs (Mostly thinking about v1 and v2 since they rely on semi-fixed seed data).

DavidTPate commented:
From what I'm seeing within [RFC 4122](https://tools.ietf.org/html/rfc4122) it wouldn't matter how these are generated as they all validate to the same style (with the exception of the version number). 

Here's the ABNF from [page 4](https://tools.ietf.org/html/rfc4122#page-4):

```
      UUID                   = time-low "-" time-mid "-"
                               time-high-and-version "-"
                               clock-seq-and-reserved
                               clock-seq-low "-" node
      time-low               = 4hexOctet
      time-mid               = 2hexOctet
      time-high-and-version  = 2hexOctet
      clock-seq-and-reserved = hexOctet
      clock-seq-low          = hexOctet
      node                   = 6hexOctet
      hexOctet               = hexDigit hexDigit
      hexDigit =
            "0" / "1" / "2" / "3" / "4" / "5" / "6" / "7" / "8" / "9" /
            "a" / "b" / "c" / "d" / "e" / "f" /
            "A" / "B" / "C" / "D" / "E" / "F"
```

DavidTPate commented on lib/string.js:
```diff
@@ -292,18 +292,56 @@ internals.String = class extends Any {
+            'uuidv5': '5'
+        };
+
+        const regex = /^([\[{\(]?)([0-9A-F]{8})([:-]?)([0-9A-F]{4})([:-]?)([0-9A-F]{4})([:-]?)([0-9A-F]{4})([:-]?)([0-9A-F]{12})([\]}\)]?)$/i;
```
We could gain some additional performance on this Regex by switching groups that we don't need the result from to non-capture groups. For example, I don't see us using the separator group `([:-]?)` so switching that to a non-capture group `(?:[:-]?)` usually gains a few hundred thousand operations/second in my tests.

DavidTPate commented on lib/string.js:
```diff
@@ -292,18 +292,56 @@ internals.String = class extends Any {
+            }
+
+            // Matching braces
+            if (brackets[results[1]] !== results[11]) {
```
I'm not sure if they are eluding me or what but I'm not seeing any tests for mismatched braces.

DavidTPate commented on lib/string.js:
```diff
@@ -292,18 +292,56 @@ internals.String = class extends Any {
 
-            return this.createError('string.guid', { value }, state, options);
+            // Matching separators
+            if (results[3] !== results[5] || results[3] !== results[7] || results[3] !== results[9]) {
```
I wasn't able to find any tests for mismatches separators (not sure if they are hiding from me).

kamronbatman commented on test/string.js:
```diff
@@ -2983,7 +2983,74 @@ describe('string', () => {
                 ['{283B67B2-430F-4E6F-97E6-19041992-C1B0}', false, null, '"value" must be a valid GUID'],
                 ['{D1A5279D-B27D-4CD4-A05E-EFDD53D08E8D', false, null, '"value" must be a valid GUID'],
-                ['D1A5279D-B27D-4CD4-A05E-EFDD53D08E8D}', false, null, '"value" must be a valid GUID']
+                ['D1A5279D-B27D-4CD4-A05E-EFDD53D08E8D}', false, null, '"value" must be a valid GUID'],
```
These are the tests I added for mismatched braces and separators.
The existing test above this (line 2985) is also a test for that.

gelliott181 replied:
Do you have tests for hyphenated UUIDs without brackets? These are technically valid also, though they require all hyphens to be included.

DavidTPate replied:
Sorry, @kamronbatman I was meaning ones like `[D1A5279D-B27D-4CD4-A05E-EFDD53D08E8D}` where the braces don't match. I definitely saw the ones your referenced above.

kamronbatman replied:
@gelliott181 There is already a test for that: line 2978.

@DavidTPate I see what you are saying and I'll add that in, thanks.

DavidTPate replied:
I would expect the tests to verify that the functionality is working as expected, not that the code is working the way it was written.

kamronbatman replied:
Ok, fair enough.

kamronbatman replied:
@gelliott181 and @DavidTPate I updated/added more tests.  Let me know what you two think.

Thanks!

DavidTPate commented:
LGTM

kamronbatman commented:
@DavidTPate I added support for string value of the versions option.  Also updated the API reference.

kamronbatman commented:
@Marsup I think this might be ready to merge! 🎉

kamronbatman commented:
@DavidTPate @gelliott181 and @Marsup - I updated the options, added assertions, and segregated the tests to follow in how the other Joi features were written (specifically I took it from IP options).  Should still be good in terms of tests/coverage.

kamronbatman commented on lib/string.js:
```diff
@@ -292,18 +292,76 @@ internals.String = class extends Any {
+            'uuidv5': '5'
+        };
+
+        let versions;
```
I was following line 186.  Not sure what you would suggest otherwise.

Marsup replied:
Oh I see, I guess the code can't always be consistent everywhere :)

kamronbatman replied:
ok, so are you thinking to move the checks to inside of the test callback? I am sorry for being confused on your thoughts.

Marsup replied:
I was suggesting a `const versions = []`, but I guess we won't be able to do the unique then, I don't think it matters much.

kamronbatman replied:
I think I changed it to what you were asking.  Let me know if that works and I will resquash.

kamronbatman commented on lib/string.js:
```diff
@@ -292,18 +292,76 @@ internals.String = class extends Any {
+
+        const regex = /^([\[{\(]?)([0-9A-F]{8})([:-]?)([0-9A-F]{4})([:-]?)([0-9A-F]{4})([:-]?)([0-9A-F]{4})([:-]?)([0-9A-F]{12})([\]}\)]?)$/i;
 
         return this._test('guid', undefined, function (value, state, options) {
```
I think I fixed it appropriately.  Let me know.

Marsup commented:
Are you able to squash all those commits as well ?

kamronbatman commented:
I just squashed it. And yes, I did not make a const because of the unique check.

DavidTPate commented:
@Marsup Not sure if you know about it or just prefer it be done differently but wanted to point out that you can now squash on merge for a PR in GitHub (you can also rebase too as of this week prior to merge). You can see the [announcement](https://github.com/blog/2141-squash-your-commits) for more details.

Marsup commented:
I know about it, they did it wrong. I want squashed commits but I still want to merge commit, none of the options work for me.

DavidTPate commented:
Roger that, @Marsup it definitely doesn't add a merge commit like you point out. Just wanted to see if it would work for you.

Marsup commented:
Thanks a lot for the work y'all, I'll release in a bit as I need to add another small thing to it.

gelliott181 commented:
Nice PR @kamronbatman, thanks for taking my minor change so much further!
